src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
-	GUS_CheckIRQ();
-	for (;;) {
-		Bit32u check=(1 << myGUS.IRQChan);
-		if (totalmask & check) return;
-		myGUS.IRQChan++;
-		if (myGUS.IRQChan>=myGUS.ActiveChannels) myGUS.IRQChan=0;
-	}
+static void GUS_DMA_Event(Bitu)
+{
+	if (gus->PerformDmaTransfer())
+		PIC_AddEvent(GUS_DMA_Event, MS_PER_DMA_XFER);
 }
 
-static Bit16u ExecuteReadRegister(void) {
-	Bit8u tmpreg;
-//	LOG_MSG("Read global reg %x",myGUS.gRegSelect);
-	switch (myGUS.gRegSelect) {
-	case 0x41: // Dma control register - read acknowledges DMA IRQ
-		tmpreg = myGUS.DMAControl & 0xbf;
-		tmpreg |= (myGUS.IRQStatus & 0x80) >> 1;
-		myGUS.IRQStatus&=0x7f;
-		return (Bit16u)(tmpreg << 8);
-	case 0x42:  // Dma address register
-		return myGUS.dmaAddr;
-	case 0x45:  // Timer control register.  Identical in operation to Adlib's timer
-		return (Bit16u)(myGUS.TimerControl << 8);
-		break;
-	case 0x49:  // Dma sample register
-		tmpreg = myGUS.DMAControl & 0xbf;
-		tmpreg |= (myGUS.IRQStatus & 0x80) >> 1;
-		return (Bit16u)(tmpreg << 8);
-	case 0x80: // Channel voice control read register
-		if (curchan) return curchan->ReadWaveCtrl() << 8;
-		else return 0x0300;
-
-	case 0x82: // Channel MSB start address register
-		if (curchan) return (Bit16u)(curchan->WaveStart >> 16);
-		else return 0x0000;
-	case 0x83: // Channel LSW start address register
-		if (curchan) return (Bit16u)(curchan->WaveStart );
-		else return 0x0000;
-
-	case 0x89: // Channel volume register
-		if (curchan) return (Bit16u)((curchan->RampVol >> RAMP_FRACT) << 4);
-		else return 0x0000;
-	case 0x8a: // Channel MSB current address register
-		if (curchan) return (Bit16u)(curchan->WaveAddr >> 16);
-		else return 0x0000;
-	case 0x8b: // Channel LSW current address register
-		if (curchan) return (Bit16u)(curchan->WaveAddr );
-		else return 0x0000;
-
-	case 0x8d: // Channel volume control register
-		if (curchan) return curchan->ReadRampCtrl() << 8;
-		else return 0x0300;
-	case 0x8f: // General channel IRQ status register
-		tmpreg=myGUS.IRQChan|0x20;
-		Bit32u mask;
-		mask=1 << myGUS.IRQChan;
-        if (!(myGUS.RampIRQ & mask)) tmpreg|=0x40;
-		if (!(myGUS.WaveIRQ & mask)) tmpreg|=0x80;
-		myGUS.RampIRQ&=~mask;
-		myGUS.WaveIRQ&=~mask;
-		CheckVoiceIrq();
-		return (Bit16u)(tmpreg << 8);
-	default:
-#if LOG_GUS
-		LOG_MSG("Read Register num 0x%x", myGUS.gRegSelect);
-#endif
-		return myGUS.gRegData;
+void Gus::StartDmaTransfers()
+{
+	PIC_AddEvent(GUS_DMA_Event, MS_PER_DMA_XFER);
+}
+
+void Gus::DmaCallback(DmaChannel *, DMAEvent event)
+{
+	if (event == DMA_UNMASKED)
+		StartDmaTransfers();
+}
+
+void Gus::PopulateAutoExec(uint16_t port, const std::string &ultradir)
+{
+	// ULTRASND=Port,(rec)DMA1,(pcm)DMA2,(play)IRQ1,(midi)IRQ2
+	std::ostringstream sndline;
+	sndline << "SET ULTRASND=" << std::hex << std::setw(3) << port << ","
+	        << std::dec << static_cast<int>(dma1) << ","
+	        << static_cast<int>(dma2) << "," << static_cast<int>(irq1)
+	        << "," << static_cast<int>(irq2) << std::ends;
```

Comment:
Since so many changes are being implemented in this PR anyway… let's remove `std::ostringstream`.

IMO, this code will be more readable if this string will be formatted using `snprintf`. It'll likely be faster as well (but it does not matter, it's not a hot code path).

Reply:
Will take a crack at it..

Reply:
Much cleaner - but MSVC added two warnings about this `snprintf` call; so had to bump the Windows allowed warning limits.